### ISSUE TITLE
Importing and using extensions does not work

### ISSUE THREAD

**Issue by kirankashalkar** [corenova/yang-forge#10]
I have the following 2 yang files:
test-ext.yang:

```
module test-ext
{
  namespace "http://test.com/ns/test-1.0/test-ext";
  prefix "testext";

  revision 2014-03-28 {
    description
      "Initial revision.";
    reference
      "Test ref";
  }

  extension msg-new {
    argument "typename";
    description
      "Just an example";
  }
}
```

test.yang:

```
module test
{
  namespace "http://test/ns/test-1.0/test";
  prefix "test";

  import test-ext {
    prefix "testext";
  }

  revision 2014-10-27 {
    description
      "Test";
    reference
      "Derived from earlier versions of base YANG files";
  }

  container testcontainer {
    testext:msg-new "TestMe";
    leaf name {
      description "Test name.";
      type string;
    }
  }
}
```

Compiling test-ext works but when I compile test.yang, it fails:

```
$ yfc schema -c test.yang -f json
error occurred during schema
Error: invalid 'msg-new' extension found during compile operation

  Usage: schema [options] [schema...]
.
.
.
```

Is this unsupported?

**Comment by sekur** [comment]
It is supported - but only by _packaging_ the schema that is exporting new extension(s).  This is because declaring/defining a new extension is essentially augmenting the language space of the compiler.  So, in order for that extension to be handled by an importing module, yangforge needs to know how it should handle that new extension. 

An example of how new modules (with extensions) can be imported by other modules can be referenced in "complex-types" and even "yang-v1-extensions" module in the /yang directory. 

You will need to do a couple of things:
1. Need to declare in package.json any overrides to define where the new extension can be used.  This is because yangforge provides sub statement context enforcement - meaning can the new extension belong in a module, a container, etc.  also need to define cardinality, can it belong multiple times, only once, etc.  
2. You will likely want to define a resolver function for yangforge to deal with that extension statement.  You need to keep in mind that yangforge is first and foremost a _compiler_.  When you declare a new extension (adding vocabulary to the compiler), it will need to know what to do when it encounters that statement.  If you do NOT define a resolver, all it will do is add that statement into the metadata section of the compiled output.  Example(s) for this is also available in the modules I've referenced above. 

Let me know if this provides enough clarity.  I suppose it's time to update the README with examples for implementing schema modules with extensions. :-)

**Comment by kirankashalkar** [comment]
Thanks.
I will try those steps but yes - examples/walkthrough for adding an
extension in the README would be super helpful.

**Comment by kirankashalkar** [comment]
Once all those steps are completed, how does one go about building the
project and installing it where npm install -g yangforge does?

**Comment by sekur** [comment]
Excellent question. :-)

Short answer is... I have yet to implement the build/publish functionality to do this easily and natively using yfc.  It's next on my todo. 

But, ultimately yfc is simply a wrapper on top of npm.  So, if you really wanted to publish it and make it installable by others, you **can** simply treat it as just another npm module and use npm publish to publish into the npm registry. 

The downside to publishing to native npm registry is that it won't be integrated into the upcoming yfc lifecycle commands, such as future planned _search and install_ features where it will query a special registry to only house YANG related modules for sharing between the community of yangforge users.  Given that this project is just getting started (about a month old), there's not much downside at this point to publish directly to npm if you choose and you can always use "yfc publish" at a future time once that facility becomes generally available. 

Peter Lee

**Comment by sekur** [comment]
@kirankashalkar were you able to follow the steps and synthesize a schema module with extensions support?  I'll close this issue for now and open another issue for improving the docs with examples for this scenario.

**Comment by kirankashalkar** [comment]
Thanks for following up.
Haven't had a look again.
OK to close this issue.
